# Milestone 2 - Machine Learning Comparative Analysis

This notebook runs the improved ML pipeline for the DSA210 project.

It compares the original project dataset with an external UCI Student Performance dataset. The external dataset is used as a separate comparative analysis, not blindly merged into the original dataset.

Main target used here: **Low / Medium / High academic performance**.

## External dataset

The added external dataset is stored as:

```text
data/external_student_performance_uci.csv
```

It combines the UCI Mathematics and Portuguese student-performance files into one course-level dataset with a `subject` column. The original `G3` final grade is transformed into `performance_level`:

```text
G3 < 10        -> Low
10 <= G3 < 15  -> Medium
G3 >= 15       -> High
```

In [ ]:
import pandas as pd

external_df = pd.read_csv('data/external_student_performance_uci.csv')
print('External dataset shape:', external_df.shape)
print('
Performance-level distribution:')
print(external_df['performance_level'].value_counts())
external_df.head()

## Run full ML pipeline

The complete implementation is in `Milestone2_ML.py`. Running the cell below executes all model comparisons and saves result tables/figures.

In [1]:
%run Milestone2_ML.py


Dataset setting: Original Higher Education Dataset - 3-Class Performance
Rows: 145
Target distribution:
Low       67
High      47
Medium    31

Model comparison:
              model  cv_accuracy_mean  cv_macro_f1_mean  test_accuracy  test_balanced_accuracy  test_macro_f1  test_weighted_f1
  Gradient Boosting             0.707             0.629          0.724                   0.682          0.682             0.729
      Decision Tree             0.533             0.480          0.621                   0.627          0.611             0.627
Logistic Regression             0.577             0.530          0.552                   0.486          0.511             0.581
         Linear SVM             0.585             0.514          0.517                   0.453          0.475             0.542
      Random Forest             0.569             0.504          0.586                   0.474          0.465             0.556
                KNN             0.551             0.463          0.55

## Read the final comparison table

In [ ]:
all_results = pd.read_csv('results_ml/all_model_comparisons.csv')
all_results[['dataset_setting', 'model', 'test_accuracy', 'test_macro_f1', 'test_weighted_f1']]    .sort_values(['dataset_setting', 'test_macro_f1'], ascending=[True, False])

## Interpretation to use in the report

The old 8-class prediction problem was too sparse because the original dataset has only 145 rows. Grouping grades into Low, Medium, and High makes the task more stable.

The external UCI dataset gives a stronger comparative ML analysis because it has 1044 course-level observations. The best external setting includes `G1` and `G2`, which are previous period grades and therefore strong predictors of `G3`. This is why the external model reaches around 0.89 test accuracy. The notebook also reports a harder early-prediction setting without `G1` and `G2`; that score is lower but more realistic when previous grades are not available.